gossip/gossip.go

```diff
@@ -283,59 +329,55 @@ func (g *Gossip) maxToleratedHops() uint32 {
 }
 
 // hasIncoming returns whether the server has an incoming gossip
-// client matching the provided address.
-func (g *Gossip) hasIncoming(addr net.Addr) bool {
+// client matching the provided node ID.
+func (g *Gossip) hasIncoming(nodeID proto.NodeID) bool {
 	g.mu.Lock()
 	defer g.mu.Unlock()
-	return g.incoming.hasAddr(addr)
+	return g.incoming.hasNode(nodeID)
+}
+
+// filterExtant removes any nodes from the supplied nodeSet which
+// are already connected to this node, either via outgoing or incoming
+// client connections.
+func (g *Gossip) filterExtant(nodes *nodeSet) *nodeSet {
+	return nodes.filter(func(a proto.NodeID) bool {
+		return !g.outgoing.hasNode(a)
+	}).filter(func(a proto.NodeID) bool {
+		return !g.incoming.hasNode(a)
+	})
 }
 
-// initializeBootstrapAddresses resolves TCP addresses supplied to the
-// constructor or set using setBootstrap(). Non-TCP and resolvable TCP
-// addresses are added to the bootstraps addrSet, except for the local
-// node, which is removed, if listed as a bootstrap host.
-func (g *Gossip) initializeBootstrapAddresses() {
+// getNextBootstrapAddress returns the next available bootstrap
+// address by consulting the first non-exhausted resolver from the
+// slice supplied to the constructor or set using setBootstrap().
+func (g *Gossip) getNextBootstrapAddress() net.Addr {
```

Comment:
not sure sure how far to follow that pattern, but should be called `getNextBootstrapAddressLocked`